numpy/core/tests/test_shape_base.py

```diff
@@ -330,5 +333,121 @@ def test_stack():
                         stack, [m, m])
 
 
+class TestBlock(TestCase):
+    def test_block_simple_row_wise(self):
+        A = np.ones((2, 2))
+        B = 2 * A
+        desired = np.array([[1, 1, 2, 2],
+                            [1, 1, 2, 2]])
+        result = block([A, B])
+        assert_equal(desired, result)
+        # with tuples
+        result = block((A, B))
+        assert_equal(desired, result)
+
+    def test_block_simple_column_wise(self):
+        A = np.ones((2, 2))
+        B = 2 * A
+        expected = np.array([[1, 1],
+                             [1, 1],
+                             [2, 2],
+                             [2, 2]])
+        result = block([[A], [B]])
+        assert_equal(expected, result)
+
+    def test_block_with_1d_arrays_row_wise(self):
+        # # # 1-D vectors are treated as row arrays
+        a = np.array([1, 2, 3])
+        b = np.array([2, 3, 4])
+        expected = np.array([1, 2, 3, 2, 3, 4])
+        result = block([a, b])
+        assert_equal(expected, result)
+
+    def test_block_with_1d_arrays_multiple_rows(self):
+        a = np.array([1, 2, 3])
+        b = np.array([2, 3, 4])
+        expected = np.array([[1, 2, 3, 2, 3, 4],
+                             [1, 2, 3, 2, 3, 4]])
+        result = block([[a, b], [a, b]])
+        assert_equal(expected, result)
+
+    def test_block_with_1d_arrays_column_wise(self):
+        # # # 1-D vectors are treated as row arrays
+        a = np.array([1, 2, 3])
+        b = np.array([2, 3, 4])
+        expected = np.array([[1, 2, 3],
+                             [2, 3, 4]])
+        result = block([[a], [b]])
+        assert_equal(expected, result)
+
+    def test_block_mixed_1d_and_2d(self):
+        A = np.ones((2, 2))
+        B = np.array([2, 2])
+        result = block([[A], [B]])
+        expected = np.array([[1, 1],
+                             [1, 1],
+                             [2, 2]])
+        assert_equal(expected, result)
+
+    def test_block_complicated(self):
+        # a bit more complicated
+        One = np.array([[1, 1, 1]])
+        Two = np.array([[2, 2, 2]])
+        Three = np.array([[3, 3, 3, 3, 3, 3]])
+        four = np.array([4, 4, 4, 4, 4, 4])
+        five = np.array([5])
+        six = np.array([6, 6, 6, 6, 6])
+        Zeros = np.zeros((2, 6))
+
+        result = block([[One, Two],
+                        Three,
+                        four,
+                        [five, six],
+                        Zeros])
+        expected = np.array([[1, 1, 1, 2, 2, 2],
```

Comment:
I would be interested to see a case similar to this, but where `Two` spans multiple rows. The reason being that none of the input arrays are actually 2D here. Something like
```
one = np.array([[1, 1, 1]])
two = np.array([[2, 2, 2], [2, 2, 2], [2, 2, 2]])
three = np.array([[3, 3, 3]])
four = np.array([[4, 4, 4]])
...
result = block([[one], [three], [four], two], [five, six], zeros)
expected = np.array([[1, 1, 1, 2, 2, 2],
                     [3, 3, 3, 2, 2, 2],
                     [4, 4, 4, 2, 2, 2],
                     [5, 6, 6, 6, 6, 6],
                     [0, 0, 0, 0, 0, 0],
                     [0, 0, 0, 0, 0, 0]])
```